pkg/apis/configuration/v1alpha1/kong_license_types.go

```diff
@@ -0,0 +1,107 @@
+	// Enabled is set to true to let controllers (like KIC) to reconcile it.
+	// Default value is true to apply the license by default.
+	// +kubebuilder:default=true
+	Enabled bool `json:"enabled"`
+	// Status is the status of the KongLicense being processed by controllers.
+	Status KongLicenseStatus `json:"status,omitempty"`
+}
+
+// KongLicenseStatus stores the status of the KongLicense being processesed in each controller that reconciles it.
+type KongLicenseStatus struct {
+	// +listType=map
+	// +listMapKey=controllerName
+	KongLicenseControllerStatuses []KongLicenseControllerStatus `json:"controllers,omitempty"`
+}
+
+// KongLicenseControllerStatus is the status of owning KongLicense being processed
+// identified by the controllerName field.
+type KongLicenseControllerStatus struct {
+	// ControllerName is an identifier of the controller to reconcile this KongLicense.
+	// Should be unique in the list of controller statuses.
+	ControllerName string `json:"controllerName"`
+	// ControllerRef is the reference of the controller to reconcile this KongLicense.
+	// It is usually the name of (KIC/KGO) pod that reconciles it.
+	ControllerRef *ControllerReference `json:"controllerRef,omitempty"`
+	// Configured is set to true if the controller applied the content of the license on managed Kong gateway.
+	Configured bool `json:"configured"`
+	// Conditions describe the current conditions of the KongLicense on the controller.
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// TODO: copy validation notes from the gateway api package to run the same validation?
+
+// Group refers to a Kubernetes Group. It must either be an empty string or a
+// RFC 1123 subdomain.
+// +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Pattern=`^$|^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+type Group string
+
+// Kind refers to a kubernetes kind.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=63
+// +kubebuilder:validation:Pattern=`^[a-zA-Z]([-a-zA-Z0-9]*[a-zA-Z0-9])?$`
+type Kind string
+
+// Namespace refers to a Kubernetes namespace. It must be a RFC 1123 label.
+// +kubebuilder:validation:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=63
+type Namespace string
+
+// ObjectName refers to the name of a Kubernetes object.
+// Object names can have a variety of forms, including RFC1123 subdomains,
+// RFC 1123 labels, or RFC 1035 labels.
+//
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=253
+type ObjectName string
+
+type ControllerReference struct {
+	// Group is the group of referent.
+	// It should be empty if the referent is in "core" group (like pod.)
+	Group *Group `json:"group,omitempty"`
+	// Kind is the kind of the referent.
+	Kind *Kind `json:"kind,omitempty"`
```

Comment:
What does an empty kind mean? Let's have this documented (I suppose it could default to `Pod`) or make it a required field.

Reply:
Done.